### PR TITLE
MULE-19326: Parameter must provide a non-empty value

### PR DESCRIPTION
**Pull request by pabloperalta**

### CODE REVIEW

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/BeanDefinitionCreator.java [hunk 1]
```diff
@@ -62,23 +61,12 @@ abstract boolean handleRequest(Map<ComponentAst, SpringComponentModel> springCom
...
-  }
-
-  protected ComponentAst resolveOwnerComponent(CreateBeanDefinitionRequest request) {
-    for (int i = request.getComponentModelHierarchy().size() - 1; i >= 0; --i) {
-      final ComponentAst possibleOwner = request.getComponentModelHierarchy().get(i);
+    return converter.map(
+                         typeConverter -> genericBeanDefinition(ConstantFactoryBean.class).addConstructorArgValue(typeConverter
+                             .convert(value)).getBeanDefinition())
+        .orElseGet(
+                   () -> genericBeanDefinition(type).addConstructorArgValue(value).getBeanDefinition());
```
Comment (pabloperalta) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/BeanDefinitionCreator.java:68]: Replaced isPresent with map/or else get

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ComponentConfigurationBuilder.java [hunk 2]
```diff
@@ -66,6 +65,7 @@
   private final ComponentAst componentModel;
   private final CreateBeanDefinitionRequest createBeanDefinitionRequest;
   private final ComponentBuildingDefinition<T> componentBuildingDefinition;
+  private final ParameterGroupUtils parameterGroupUtils = new ParameterGroupUtils();
```
Comment (pabloperalta) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ComponentConfigurationBuilder.java:68]: Made final

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtils.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+  /**
+   * A parameter corresponding to a dsl syntax element maybe nested inside parameter groups that will not show in the dsl. Because
+   * of this, in order to find the entry corresponding to a dsl parameter name in the generation information, a tree-like
+   * structure must be traversed. A breadth search is used because the most common case is that parameters are in the first level.
+   *
+   * @param name:             The dsl name
+   * @param dslElementSyntax: The root syntax with the generation information
+   * @return The corresponding parameter name to a dsl name
+   */
```
Comment (pabloperalta) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtils.java:73]: Added Javadocs

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtils.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    return null;
+  }
+
+  private List<Pair<DslElementSyntax, String>> makeElementsToNamePairList(DslElementSyntax dslElementSyntax) {
+    return dslElementSyntax.getContainedElementsByName().entrySet().stream()
+        .map(entry -> new Pair<>(entry.getValue(), entry.getKey()))
+        .collect(Collectors.toList());
+  }
+
```
Comment (pabloperalta) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtils.java:95]: Used standard Pair implementation

File: modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtilsTestCase.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.config.internal.dsl.spring;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.nullValue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
```
Comment (pabloperalta) [inline modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtilsTestCase.java:13]: Moved static imports to the top

File: modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/SimpleTypeBeanDefinitionCreatorTestCase.java [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+ */
+package org.mule.runtime.config.internal.dsl.spring;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
```
Comment (pabloperalta) [inline modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/SimpleTypeBeanDefinitionCreatorTestCase.java:15]: Moved static imports to the top

File: modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtilsTestCase.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+import org.mule.runtime.ast.api.ComponentAst;
+import org.mule.runtime.ast.api.ComponentGenerationInformation;
+import org.mule.runtime.ast.api.ComponentMetadataAst;
+import org.mule.runtime.ast.api.ComponentParameterAst;
+import org.mule.runtime.ast.internal.DefaultComponentAst;
+import org.mule.runtime.ast.internal.builder.PropertiesResolver;
+import org.mule.runtime.extension.api.dsl.syntax.DslElementSyntax;
+import org.mule.tck.junit4.AbstractMuleTestCase;
+
+public class ParameterUtilsTestCase extends AbstractMuleTestCase {
```
Comment (pabloperalta) [inline modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtilsTestCase.java:42]: Extended AbstractMuleTestCase

File: modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/SimpleTypeBeanDefinitionCreatorTestCase.java [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+import org.mule.runtime.ast.api.ComponentMetadataAst;
+import org.mule.runtime.ast.api.ComponentParameterAst;
+import org.mule.runtime.config.internal.dsl.model.SpringComponentModel;
+import org.mule.runtime.config.internal.dsl.processor.ObjectTypeVisitor;
+import org.mule.runtime.dsl.api.component.ComponentBuildingDefinition;
+import org.mule.runtime.dsl.api.component.TypeConverter;
+import org.mule.tck.junit4.AbstractMuleTestCase;
+import org.springframework.beans.factory.config.BeanDefinition;
+
+public class SimpleTypeBeanDefinitionCreatorTestCase extends AbstractMuleTestCase {
```
Comment (pabloperalta) [inline modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/SimpleTypeBeanDefinitionCreatorTestCase.java:35]: Extended AbstractMuleTestCase

File: modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtilsTestCase.java [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    ComponentMetadataAst metadata = null;
+
+    componentAst =
+        new DefaultComponentAst(builtChildren, parameterAsts, extraParams, propertiesResolver, componentId, extensionModel,
+                                componentModel, nestableElementModel, configurationModel, connectionProviderModel,
+                                parameterizedModel, generationInformation, metadata, location, identifier, componentType);
+  }
+
+  @Test
+  @Issue("MULE-19326")
```
Comment (pabloperalta) [inline modules/spring-config/src/test/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtilsTestCase.java:76]: Created an Issue in MULE project in order to link the tests

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtils.java [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+      queue.addAll(makeElementsToNamePairList(currentNode.getFirst()));
+    }
+
+    return null;
+  }
+
+  private List<Pair<DslElementSyntax, String>> makeElementsToNamePairList(DslElementSyntax dslElementSyntax) {
+    return dslElementSyntax.getContainedElementsByName().entrySet().stream()
+        .map(entry -> new Pair<>(entry.getValue(), entry.getKey()))
+        .collect(Collectors.toList());
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtils.java:93]: static import

File: modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtils.java [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+        return currentNode.getSecond();
+      }
+
+      queue.addAll(makeElementsToNamePairList(currentNode.getFirst()));
+    }
+
+    return null;
+  }
+
+  private List<Pair<DslElementSyntax, String>> makeElementsToNamePairList(DslElementSyntax dslElementSyntax) {
```
Comment (elrodro83) [inline modules/spring-config/src/main/java/org/mule/runtime/config/internal/dsl/spring/ParameterUtils.java:90]: add comments to this code

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Review by elrodro83 — APPROVED** [review]

**Pull request merged**